### PR TITLE
Metamodule

### PR DESCRIPTION
**Pull request by remmeier**
exposes all kinds of type information for:
- resources
- meta information
- links information
- enums

for each of those it includes:
- attributes with type information
- super types
- sub types
- primary key info
- ...

Still considered to be very experimental.

Implemented as new module MetaModule in katharsis-meta. Most code obtained from katharsis-jpa that did make use of such a meta model in the past.

### CODE REVIEW

**Comment by remmeier** [comment]
not ready yet

**Comment by remmeier** [comment]
is there no coverage anymore?

**Comment by remmeier** [comment]
without coverage it can also be merged, I will do an other coverage anyway in the near future.

File: katharsis-jpa/src/main/java/io/katharsis/jpa/meta/internal/AbstractJpaDataObjectProvider.java [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+		if (element.getParent() instanceof MetaJpaDataObject && element instanceof MetaAttribute) {
+			MetaAttribute attr = (MetaAttribute) element;
+			MetaDataObject parent = attr.getParent();
+			Type implementationType = PropertyUtils.getPropertyType(parent.getImplementationClass(), attr.getName());
+
+			Class<?> elementType = getElementType(implementationType);
+
+			boolean jpaObject = attr.isAssociation() || elementType.getAnnotation(Embeddable.class) != null;
+
+			MetaType metaType = context.getLookup().getMeta(implementationType, jpaObject ? MetaJpaDataObject.class : MetaType.class);
```
Comment (chb0github) [inline katharsis-jpa/src/main/java/io/katharsis/jpa/meta/internal/AbstractJpaDataObjectProvider.java]: I am trying to do the build and release and this does not compile: 
```
ERROR] /Users/cbongiorno/development/katharsis-framework/katharsis-jpa/src/main/java/io/katharsis/jpa/meta/internal/AbstractJpaDataObjectProvider.java:[33,64] no suitable method found for getMeta(java.lang.reflect.Type,jpaObject [...]class)
[ERROR] method io.katharsis.meta.MetaLookup.<T>getMeta(java.lang.reflect.Type,java.lang.Class<T>) is not applicable
[ERROR] (inferred type does not conform to equality constraint(s)
[ERROR] inferred: io.katharsis.meta.model.MetaType
[ERROR] equality constraints(s): io.katharsis.meta.model.MetaType,io.katharsis.jpa.meta.MetaJpaDataObject)
[ERROR] method io.katharsis.meta.MetaLookup.<T>getMeta(java.lang.reflect.Type,java.lang.Class<T>,boolean) is not applicable
[ERROR] (cannot infer type-variable(s) T
[ERROR] (actual and formal argument lists differ in length))
[ERROR] /Users/cbongiorno/development/katharsis-framework/katharsis-jpa/src/main/java/io/katharsis/jpa/JpaModule.java:[420,49] no suitable method found for getMeta(java.lang.Class<capture#1 of ?>,mapped ? M[...]class)
[ERROR] method io.katharsis.meta.MetaLookup.<T>getMeta(java.lang.reflect.Type,java.lang.Class<T>) is not applicable
[ERROR] (inferred type does not conform to equality constraint(s)
[ERROR] inferred: io.katharsis.jpa.meta.MetaJpaDataObject
[ERROR] equality constraints(s): io.katharsis.jpa.meta.MetaJpaDataObject,io.katharsis.meta.model.resource.MetaJsonObject)
[ERROR] method io.katharsis.meta.MetaLookup.<T>getMeta(java.lang.reflect.Type,java.lang.Class<T>,boolean) is not applicable
[ERROR] (cannot infer type-variable(s) T
[ERROR] (actual and formal argument lists differ in length))

```

**Pull request merged**